soc/arm/atmel_sam/common/soc_pmc.c

```diff
@@ -68,3 +70,317 @@ uint32_t soc_pmc_peripheral_is_enabled(uint32_t id)
 	}
 	return 0;
 }
+
+void soc_pmc_mck_set_prescaler(uint32_t prescaler)
+{
+	uint32_t reg_val;
+
+	switch (prescaler) {
+	case 1:
+		reg_val = PMC_MCKR_PRES_CLK_1;
+		break;
+	case 2:
+		reg_val = PMC_MCKR_PRES_CLK_2;
+		break;
+	case 4:
+		reg_val = PMC_MCKR_PRES_CLK_4;
+		break;
+	case 8:
+		reg_val = PMC_MCKR_PRES_CLK_8;
+		break;
+	case 16:
+		reg_val = PMC_MCKR_PRES_CLK_16;
+		break;
+	case 32:
+		reg_val = PMC_MCKR_PRES_CLK_32;
+		break;
+	case 64:
+		reg_val = PMC_MCKR_PRES_CLK_64;
+		break;
+	case 3:
+		reg_val = PMC_MCKR_PRES_CLK_3;
+		break;
+	default:
+		__ASSERT(false, "Invalid MCK prescaler");
+		reg_val = PMC_MCKR_PRES_CLK_1;
+		break;
+	}
+
+	PMC->PMC_MCKR = (PMC->PMC_MCKR & (~PMC_MCKR_PRES_Msk)) | reg_val;
+
+	while (!(PMC->PMC_SR & PMC_SR_MCKRDY)) {
+	}
+}
+
+#if defined(CONFIG_SOC_SERIES_SAME70) || defined(CONFIG_SOC_SERIES_SAMV71)
+void soc_pmc_mck_set_division(uint32_t divider)
+{
+	uint32_t reg_val;
+
+	switch (divider) {
+	case 1:
+		reg_val = PMC_MCKR_MDIV_EQ_PCK;
+		break;
+	case 2:
+		reg_val = PMC_MCKR_MDIV_PCK_DIV2;
+		break;
+	case 3:
+		reg_val = PMC_MCKR_MDIV_PCK_DIV3;
+		break;
+	case 4:
+		reg_val = PMC_MCKR_MDIV_PCK_DIV4;
+		break;
+	default:
+		__ASSERT(false, "Invalid MCK divider");
+		reg_val = PMC_MCKR_MDIV_EQ_PCK;
+		break;
+	}
+
+	PMC->PMC_MCKR = (PMC->PMC_MCKR & (~PMC_MCKR_MDIV_Msk)) | reg_val;
+
+	while (!(PMC->PMC_SR & PMC_SR_MCKRDY)) {
+	}
+}
+#endif /* CONFIG_SOC_SERIES_SAME70 || CONFIG_SOC_SERIES_SAMV71 */
+
+void soc_pmc_mck_set_source(enum soc_pmc_mck_src source)
```

Comment:
I think makes more sense set the prescaler together at _soc_pmc_mck_set_source(source, prescaler);_

Reply:
I start to believe that this will be the best signature because we don't know what is the current config. So, when switching the master clock it is better already reinforce what it is the prescaler.

Reply:
@nandojve,

Problem with this, is that the `same70` and `samv71` have both a prescaler and a divider. So the function signature would be different.

How do you prefer that this is handled?

Reply:
I've merged `set_source` with prescaler and divider, where the divider is only used on supported SoCs.